### <center>Algoritmos e Estruturas de Dados</center>

#### <center>Números reais</center>


**Prof.**: João Paulo Orlando

**Curso**: Técnico em Informática Integrado ao Ensino Médio

Os números de ponto flutuante são representados no hardware do computador como frações de base 2 (binárias). Por exemplo, a fração decimal 0.125 tem valor 1/10 + 2/100 + 5/1000.

Infelizmente, muitas frações decimais não podem ser representadas precisamente como frações binárias. O resultado é que, em geral, os números decimais de ponto flutuante que você digita acabam sendo armazenados de forma apenas aproximada, na forma de números binários de ponto flutuante.

O problema é mais fácil de entender primeiro em base 10. Considere a fração 1/3. Podemos representá-la aproximadamente como uma fração base 10:

`0.3`

ou melhor,

`0.33`

ou melhor,

`0.333`

e assim por diante. Não importa quantos dígitos você está disposto a escrever, o resultado nunca será exatamente `1/3`, mas será uma aproximação de cada vez melhor de `1/3`.

Da mesma forma, não importa quantos dígitos de base 2 estejas disposto a usar, o valor decimal 0.1 não pode ser representado exatamente como uma fração de base 2. No sistema de base 2, `1/10` é uma fração binária que se repete infinitamente:

`0.0001100110011001100110011001100110011001100110011...`

Se parares em qualquer número finito de bits, obterás uma aproximação. Hoje em dia, na maioria dos computadores, as casas decimais são aproximados usando uma fração binária onde o numerado utiliza os primeiros 53 bits iniciando no bit mais significativo e tendo como denominador uma potência de dois. No caso de `1/10`, a fração binária seria `602879701896397 / 2 ** 55` o que chega bem perto, mas mesmo assim, não é igual ao valor original de `1/10`.

É fácil esquecer que o valor armazenado é uma aproximação da fração decimal original, devido à forma como os floats são exibidos no interpretador interativo. O Python exibe apenas uma aproximação decimal do verdadeiro valor decimal da aproximação binária armazenada pela máquina. Se o Python exibisse o verdadeiro valor decimal da aproximação binária que representa o decimal 0.1, seria necessário mostrar:

`
0.1
0.1000000000000000055511151231257827021181583404541015625`

Contém muito mais dígitos do que é o esperado e utilizado pela grande maioria dos desenvolvedores, portanto, o Python limita o número de dígitos exibidos, apresentando um valor arredondado, ao invés de mostrar todas as casas decimais:

`
1 / 10
0.1`

Lembre-se, mesmo que o resultado impresso seja o valor exato de 1/10, o valor que verdadeiramente estará armazenado será uma fração binária representável que mais se aproxima.

In [5]:
print(1/3)

0.3333333333333333


In [20]:
print(f"{1/3:.100f}")

0.3333333333333333148296162562473909929394721984863281250000000000000000000000000000000000000000000000


In [39]:
# Importa a biblioteca com funções prontas. Estamos usando a função que calcula o cosseno com math.cos()
import math

x = 0.5

print(f'cos({x}) = {math.cos(x)}')

cos(0.5) = 0.8775825618903728


Podemos modificar o número de casas decimais da mensagem de saída, repare na diferença para o anterior. Adicionamos ```:.1f``` e ```:.3f``` entre as chaves.

In [10]:
import math

x = 0.5

print(f'cos({x}) = {math.cos(x):.1f}')
print(f'cos({x}) = {math.cos(x):.3f}')

cos(0.5) = 0.9
cos(0.5) = 0.878


Por enquanto, o que vamos precisar saber, é somente formatar as casas decimais da saída. Vários exercícios irão solicitar isso de vocês.

Entretanto, a formatação possui diversos recursos bem interessantes. Para saber mais, consulte a [documentação](https://docs.python.org/3/library/string.html#format-specification-mini-language).

In [14]:
# Faça um programa que leia 2 valores float (números reais) e exiba as 4 operações básicas.
#    mostrando os valores sempre com 4 casas decimais

valor1 = float(input())
valor2 = float(input())

soma = valor1 + valor2
sub = valor1 - valor2
mult = valor1 * valor2
div = valor1 / valor2

print(f"Soma = {soma:.4f}")
print(f"Subtração = {sub:.4f}")
print(f"Multiplicação = {mult:.4f}")
print(f"Divisão = {div:.4f}")

12.45
74.2
Soma = 86.6500
Subtração = -61.7500
Multiplicação = 923.7900
Divisão = 0.1678


In [19]:
# Faça um programa que leia 3 notas numerada de 0.0 até 10.0, 
#    calcule e exiba a média com 2 casas decimais.

nota1 = float(input())
nota2 = float(input())
nota3 = float(input())

media =  (nota1+nota2+nota3)/3

print(f"A média do aluno foi {media:.2f}")

10
7
0
A média do aluno foi 5.67


In [18]:
#Mesma coisa do programa anterior, mas agora calcule a média ponderada, 
#    onde as notas tem os seguintes pesos: 25, 30 e 45.

nota1 = float(input())
nota2 = float(input())
nota3 = float(input())

media =  (nota1*25+nota2*30+nota3*45)/100

print(f"A média do aluno foi {media:.2f}")

10
7
0
A média do aluno foi 4.60


In [21]:
# Faça um programa em python que calcule a área de um triângulo
#   dado a base e a altura em números float. 
#   Imprima a área com 3 casas decimais.

###   A = b*h/2    ###

base = float(input())
altura = float(input())

area = (base * altura)/2

print(f"A área é {area:.3f}")


3.33
6.4433
A área é 10.728


In [22]:
# Faça um programa em python que leia um grau em Celsius e converta ele para Fahrenheit.
#     Exiba o valores com duas casas decimais.

###    F = (C * 9/5) + 32  ###

celsius = float(input())
fahr = (celsius * 9/5) + 32
print(f"O grau {celsius} em celsius corresponde a {fahr:.2f} em Fahrenheit.")

30
O grau 30.0 em celsius corresponde a 86.00 em Fahrenheit.


In [3]:
# Faça um programa que calcule o IMC de uma pessoa.
# Leia o Nome, Peso e Altura da pessoa.
# Aplique a fórmula:        IMC = peso/(altura**2)
# Exiba o resultado com 1 casa decimal.

nome = input()
peso = float(input())
altura = float(input())

imc = peso/(altura**2)
print(f"{nome} possui imc = {imc:.1f}")

João
89
1.85
João possui imc = 26.0


Vamos analisar uma situação peculiar de alguns problemas que vocês irão resolver de agora em diante. Observe as entradas do problema 1013.

<img src="./Imagens/05-01.png">

O método `input()` lê uma linha inteira de informações, independente do que está escrito lá.

Vamos simular a entrada desse problema:

In [4]:
entrada = input()
print(f"""A entrada foi '{entrada}'""")

7 14 106
A entrada foi '7 14 106'


O que é a informação '7 14 106'?

É somente um texto, precisamos separar esse texto, exatamente em 3 partes. 

Para isso, existe um método chamado `split()` que corta o texto pelos espaços em branco. Esse método será aplicado após a execução do `input()`.

In [8]:
valorA, valorB, valorC = input().split()
print(f"""Agora temos 3 valores: '{valorA}', '{valorB}' e '{valorC}'""")


7 14 106
Agora temos 3 valores: '7', '14' e '106'


Entretanto, temos mais problema, precisamos converter cada um dos elementos para seu tipo adequado, pois eles são somente texto e não números.

In [9]:
print(type(valorA))

<class 'str'>


In [10]:
valorA = int(valorA)
print(type(valorA))

<class 'int'>


Então a leitura desses 3 valores de entrada na mesma linha poderia ficar algo assim.


In [16]:
valorA, valorB, valorC = input().split()
valorA = int(valorA)
valorB = int(valorB)
valorC = int(valorC)

#Podendo continuar o desenvolvimento de assim por diante.

maiorAB = (valorA + valorB + abs(valorA - valorB))/2

maiorABC = (maiorAB + valorC + abs(maiorAB - valorC))/2

print(f"{int(maiorABC)} eh o maior")


7 14 106
106 eh o maior


Agora podem começar a fazer a **Lista 3 - Números reais**.

**Dica**: A quantidade de linhas de entrada, será a quantidade de `input()` que deverá ser feito no seu programa.
